### PR TITLE
JIRA: AOAF-45

### PR DESCRIPTION
**Pull request by ch-skathirvel**
Description : Added payment service, router and handler to do payment.

### CODE REVIEW

File: src/services/external/payment.service.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
+import { Session } from "@companieshouse/node-session-handler";
+import ApiClient from "@companieshouse/api-sdk-node/dist/client";
+import { createPaymentApiClient } from "../internal/api.client.service";
+import { CreatePaymentRequest, Payment } from "@companieshouse/api-sdk-node/dist/services/payment";
+import { v4 as uuidv4 } from "uuid";
+import { createAndLogError, logger } from "../../utils/logger";
+import { ApiResponse } from "@companieshouse/api-sdk-node/dist/services/resource";
+import { getPaymentResourceUri, PAYMENT_REDIRECT_URI } from "../../utils/url";
+
+export const startPaymentsSession = async (session: Session, paymentSessionUrl: string, accountsFilingId: string, transactionId: string): Promise<ApiResponse<Payment>> => {
```
Comment (amartin7663) [inline src/services/external/payment.service.ts:10]: This file is formatted with 2 space tabs, but should be 4

File: src/services/external/payment.service.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  const apiClient: ApiClient = createPaymentApiClient(session, paymentSessionUrl);
+  const resourceWithHost = getPaymentResourceUri(transactionId);
+  const reference: string = "Accounts_Filing_" + accountsFilingId;
```
Comment (amartin7663) [inline src/services/external/payment.service.ts:13]: I think the reference should be Package Accounts, as that's more specific

File: src/services/external/payment.service.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  const redirectUri = PAYMENT_REDIRECT_URI;
```
Comment (amartin7663) [inline src/services/external/payment.service.ts:14]: Could you just pass this straight into the request object

File: src/services/external/payment.service.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  if (paymentResult.isFailure()) {
+    const errorResponse = paymentResult.value;
+    logger.error(`payment.service failure to create payment - http response status code = ${errorResponse?.httpStatusCode} - ${JSON.stringify(errorResponse?.errors)}`);
+    if (errorResponse.httpStatusCode === 401 || errorResponse.httpStatusCode === 429) {
+      return Promise.reject(createAndLogError(`payment.service Http status code ${errorResponse.httpStatusCode} - Failed to create payment,  ${JSON.stringify(errorResponse?.errors) || "Unknown Error"}`));
+    } else {
+      return Promise.reject(createAndLogError(`payment.service Unknown Error ${JSON.stringify(errorResponse?.errors) || "No Errors found in response"}`));
+    }
+  } else {
+    logger.info(`Create payment, status_code=${paymentResult.value.httpStatusCode}`);
```
Comment (amartin7663) [inline src/services/external/payment.service.ts:36]: Not sure you need to include the status code here, just a log message to say the payment session has been created

File: src/routers/handlers/payment_callback/payment.callback.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+        const paymentStatus = req.query.status;
+        const accountsFilingId = must(getAccountsFilingId(req.session));
+        const returnedState = req.query.state;
+        const savedState = session.getExtraData("payment-nonce");
+        if (!savedState || savedState !== returnedState) {
+          throw createAndLogError(`Returned state: ${returnedState}, saved state: ${savedState}`);
+        }
+    
+        if (paymentStatus === "paid") {
+          logger.info("Accounts filing id: " + accountsFilingId + " - Payment status: " + paymentStatus + " - redirecting to the submission page");
```
Comment (amartin7663) [inline src/routers/handlers/payment_callback/payment.callback.ts:30]: These strings would be cleaner using `
```suggestion
          logger.info(`Accounts filing id: ${accountsFilingId} - Payment status: ${paymentStatus} - redirecting to the submission page`);
```

File: src/routers/handlers/payment_callback/payment.callback.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
+import { Request, Response } from "express";
+import { GenericHandler } from "./../generic";
+import { createAndLogError, logger } from "../../../utils/logger";
+import { PrefixedUrls } from "../../../utils/constants/urls";
+import { must } from "../../../utils/session";
+import { getAccountsFilingId } from "../../../utils/session";
+import { Session } from "@companieshouse/node-session-handler";
+
+export class PaymentCallbackHandler extends GenericHandler {
```
Comment (amartin7663) [inline src/routers/handlers/payment_callback/payment.callback.ts:9]: Again this page seems to be using 2 space tabs

File: src/routers/handlers/payment_callback/payment.callback.ts [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+        super({
+            title: "Payment Callback – GOV.UK",
+            backURL: null,
+        });
+    }
+
+    async executeGet(
+      req: Request,
+      _response: Response) {
+        const session = req.session as Session;
```
Comment (amartin7663) [inline src/routers/handlers/payment_callback/payment.callback.ts:20]: Could we just use req.session in places we need it?

File: src/routers/handlers/payment_callback/payment.callback.ts [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+        const paymentStatus = req.query.status;
+        const accountsFilingId = must(getAccountsFilingId(req.session));
```
Comment (amartin7663) [inline src/routers/handlers/payment_callback/payment.callback.ts:22]: Do we need the accounts filing id, if it's only for the log message, the log messages context id should be able to identify it

File: src/routers/handlers/check_your_answers/check.your.answers.ts [hunk 9]
```diff
@@ -47,9 +51,23 @@ export class CheckYourAnswersHandler extends GenericHandler {
...
 
-        return PrefixedUrls.CONFIRMATION;
+        if (!paymentUrl) {
+            logger.debug(`No payment url ${paymentUrl} from closeTransaction, journey redirected to confirmation page`);
+            return PrefixedUrls.CONFIRMATION;
+        } else {
+            // Payment required, start the payment journey
+            logger.debug(`Received payment url ${paymentUrl} from closeTransaction, payment journey started`);
+            const paymentResponse: ApiResponse<Payment> = await startPaymentsSession(req.session, paymentUrl,
+                                                                                     must(getAccountsFilingId(req.session)), must(getTransactionId(req.session)));
```
Comment (amartin7663) [inline src/routers/handlers/check_your_answers/check.your.answers.ts:63]: Not sure I like this format, maybe
```suggestion
            const paymentResponse: ApiResponse<Payment> = await startPaymentsSession(req.session, paymentUrl,
                    must(getAccountsFilingId(req.session)), must(getTransactionId(req.session)));
```

File: src/services/external/payment.service.ts [hunk 10]
```diff
@@ -0,0 +1,40 @@
+import { Session } from "@companieshouse/node-session-handler";
+import ApiClient from "@companieshouse/api-sdk-node/dist/client";
+import { createPaymentApiClient } from "../internal/api.client.service";
+import { CreatePaymentRequest, Payment } from "@companieshouse/api-sdk-node/dist/services/payment";
+import { v4 as uuidv4 } from "uuid";
+import { createAndLogError, logger } from "../../utils/logger";
+import { ApiResponse } from "@companieshouse/api-sdk-node/dist/services/resource";
+import { getPaymentResourceUri, PAYMENT_REDIRECT_URI } from "../../utils/url";
+
+export const startPaymentsSession = async (session: Session, paymentSessionUrl: string, accountsFilingId: string, transactionId: string): Promise<ApiResponse<Payment>> => {
```
Comment (amartin7663) [inline src/services/external/payment.service.ts:10]: This line is a bit long, could we shorten it

File: src/routers/handlers/check_your_answers/check.your.answers.ts (hunk 9 above)
Comment (ch-skathirvel) [reply src/routers/handlers/check_your_answers/check.your.answers.ts:63]: Are you OK with the below format?

**Review by amartin7663 — APPROVED** [review]

**Pull request merged**